[KAT-1006, KAT-1007] logging: add first iteration of tracer logic

patrickkenney9801 opened the pull request:
Add tracer logic for Katana and a basic Json tracer for testing
This work is based on RFC-034

This code is still in development and many parts are intentionally left underdeveloped to help testing and debugging the infrastructure.  The next step is to test this on distributed workloads and then add an OpenTracer backend to replace JsonTracer.

l-hoang commented on libcusp/include/galois/graphs/Distribution.h:
```diff
@@ -0,0 +1,718 @@
+    increment_evilPhase();
+  }
+
+  //! Increments evilPhase, a phase counter used by communication.
```
I would prefer if this (and anything else related to explicit communication) be moved to the Partitioner file.

Reasoning is that in my view a Distribution object shouldn't have to be doing communication in order to figure out master/mirror assignment; that should be the burden of the Partitioner to do and save the info into the Distribution object.

l-hoang commented on libcusp/include/galois/graphs/Partitioner.h:
```diff
@@ -748,7 +402,7 @@ class Partitioner : public DistGraph<NodeTy, EdgeDataTy> {
         // note the +1 on evil phase; load messages send using a different
         // phase to avoid conflicts
-        net.sendTagged(h, base_DistGraph::evilPhasePlus1(), b);
+        net.sendTagged(h, dist_.evilPhasePlus1(), b);
```
Related to the comment about Dist maintaining evilphase; this dependence on the Distribution object for maintaining communication info is something I want to avoid.

gurbinder533 commented on libcusp/include/galois/graphs/DistGraph.h:
```diff
@@ -718,7 +108,7 @@ class DistGraph {
    * @returns Local ID of destination of edge ni
    */
   GraphNode getEdgeDst(edge_iterator ni) {
```
We need to sanitize the function names. It is mixed at the moment: camelCase and snake_case, some of which is my doing. 
It doesn't have to be part of this PR though.

ddn0 commented:
Filed an issue to continue the separation of Distribution and Partitioner(s) #377

patrickkenney9801 committed:
logging: add first iteration of tracer logic

add python script for assembling traces to be compatible
with Jaeger's OpenTracing visualization software

patrickkenney9801 commented:
Example output from the test written:

{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"2","span_name":"first span","parent_id":"null"},"log":{"msg":"start","timestamp_us":1627316629070870,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"host_data":{"hosts":1,"hostname":"PK-Desktop","hardware_threads":32,"ram_gb":100}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"2"},"tags":[{"name":"life","value":42},{"name":"type","value":"test"},{"name":"real","value":false},{"name":"somethin","value":2.000000},{"name":"hello","value":"world"}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"3","span_name":"reading query list","parent_id":"2"},"log":{"msg":"start","timestamp_us":1627316629070937,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"3"},"log":{"msg":"query parsed","timestamp_us":1627316629070953,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"tags":[{"name":"query","value":"CREATE"},{"name":"ops","value":4}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"4","span_name":"running query","parent_id":"3"},"log":{"msg":"start","timestamp_us":1627316629070973,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"5","span_name":"writing query results","parent_id":"4"},"log":{"msg":"start","timestamp_us":1627316629070993,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"6","span_name":"passing error","parent_id":"5"},"log":{"msg":"start","timestamp_us":1627316629071011,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"6","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071029,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"7","span_name":"getting error","parent_id":"5"},"log":{"msg":"start","timestamp_us":1627316629071048,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"7","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071065,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"5"},"log":{"msg":"error writing","timestamp_us":1627316629071089,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"tags":[{"name":"event","value":"error"},{"name":"error.kind","value":"arrow error"},{"name":"error.context","value":"passed along by GetError: failed to make fixed size type (tracing.cpp:6)"}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"5"},"tags":[{"name":"error","value":true}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"5","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071111,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"4","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071129,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"8","span_name":"running query","parent_id":"3"},"log":{"msg":"start","timestamp_us":1627316629071147,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"9","span_name":"writing query results","parent_id":"8"},"log":{"msg":"start","timestamp_us":1627316629071165,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"10","span_name":"passing error","parent_id":"9"},"log":{"msg":"start","timestamp_us":1627316629071183,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"10","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071200,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"11","span_name":"getting error","parent_id":"9"},"log":{"msg":"start","timestamp_us":1627316629071218,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"11","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071236,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"9"},"log":{"msg":"error writing","timestamp_us":1627316629071255,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"tags":[{"name":"event","value":"error"},{"name":"error.kind","value":"arrow error"},{"name":"error.context","value":"passed along by GetError: failed to make fixed size type (tracing.cpp:6)"}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"9"},"tags":[{"name":"error","value":true}]}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"9","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071274,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"8","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071292,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"12","span_name":"is not initially active","parent_id":"3"},"log":{"msg":"start","timestamp_us":1627316629071310,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"13","span_name":"no raii","parent_id":"12"},"log":{"msg":"start","timestamp_us":1627316629071328,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"13"},"log":{"msg":"no raii is the active span","timestamp_us":1627316629071346,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"13","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071363,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"12","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071380,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"3","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071398,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"1","host":0,"offset_ms":0,"span_data":{"span_id":"2","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071415,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"15","span_name":"first span of second trace","parent_id":"null"},"log":{"msg":"start","timestamp_us":1627316629071434,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"host_data":{"hosts":1,"hostname":"PK-Desktop","hardware_threads":32,"ram_gb":100}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"15"},"log":{"msg":"testing contexts","timestamp_us":1627316629071453,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"tags":[{"name":"trace_id","value":"14"},{"name":"span_id","value":"15"}]}
{"trace_id":"16","host":0,"offset_ms":0,"span_data":{"span_id":"17","span_name":"root span of new trace","parent_id":"null"},"log":{"msg":"start","timestamp_us":1627316629071473,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"host_data":{"hosts":1,"hostname":"PK-Desktop","hardware_threads":32,"ram_gb":100}}
{"trace_id":"16","host":0,"offset_ms":0,"span_data":{"span_id":"17"},"log":{"msg":"the new root span of trace 3 is active","timestamp_us":1627316629071490,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"18","span_name":"child of trace 2's root by context","parent_id":"15"},"log":{"msg":"start","timestamp_us":1627316629071509,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000},"host_data":{"hosts":1,"hostname":"PK-Desktop","hardware_threads":32,"ram_gb":100}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"18"},"log":{"msg":"child span of trace 2 is active","timestamp_us":1627316629071527,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"18","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071544,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"16","host":0,"offset_ms":0,"span_data":{"span_id":"17","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071562,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}
{"trace_id":"14","host":0,"offset_ms":0,"span_data":{"span_id":"15","finished":true},"log":{"msg":"finished","timestamp_us":1627316629071578,"max_mem_gb":0.024,"mem_gb":0.024,"arrow_mem_gb":0.000}}

ddn0 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -0,0 +1,132 @@
+#include "katana/ProgressTracer.h"
+
+namespace katana {
```
Can we remove this `namespace katana` line and just prefix symbols with `katana` explicitly?

You'll only need to add `katana::` to function names and parameters. Once the parser knows the function is in the katana namespace (e.g., the body of the function), unqualified names will be searched in the katana namespace.

ddn0 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -0,0 +1,132 @@
+  }
+}
+
+/// Create a new top level span if one does not already exist.
```
These doxygen comments probably should be in the prototype defined in the header file.

ddn0 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -0,0 +1,132 @@
+  }
+}
+
+/// Closes the underlying ProgressSpan if the ProgressScope was created
```
ditto

ddn0 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -0,0 +1,132 @@
+  return SetActiveSpan(StartSpan(span_name, child_of), finish_on_close);
+}
+
+/// Set the current active thread-local span to provided span
```
ditto

ddn0 commented on libsupport/src/ProgressTracer.cpp:
```diff
@@ -0,0 +1,132 @@
+  return scope_closed_;
+}
+
+/// Every ProgressSpan started must be finished
```
ditto

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+
+namespace katana {
+
+class KATANA_EXPORT JsonTracer : public ProgressTracer {
```
"JSONTracer"

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+  static std::shared_ptr<JsonTracer> Make(
+      uint32_t host_id = 0, uint32_t num_hosts = 1);
+
+  /// Create a new top level span if ignore_active_span=true and
```
Unless `JsonTracer::StartSpan` has caveats that aren't covered by `ProgressTracer::StartSpan`, there isn't much need to copy the same documentation again.

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+  /// Otherwise creates a child span of the child_of span or active span
+  /// Should only be used to handle multiple active spans simultaneously
+  std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, bool ignore_active_span = false) override;
```
Virtual functions and default parameters do not mix very well. If we really want this API, it should be written like:

```
class ProgressLogger:
public:
virtual std::shared_ptr<ProgressSpan> StartSpan(
      const std::string& span_name, bool ignore_active_span);

std::shared_ptr<ProgressSpan> StartSpan(const std::string& span_name) {
  return StartSpan(span_name, false);
}
};
```

ddn0 commented on libgalois/src/SharedMemSys.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "katana/SharedMemSys.h"
 
 #include "katana/CommBackend.h"
+#include "katana/JsonTracer.h"
```
"katana/JSONTracer.h"

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+
+public:
+  std::unique_ptr<ProgressContext> Clone() const noexcept override;
+  std::string GetTraceId() const noexcept override { return trace_id_; }
```
"GetTraceID", etc.

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+      : trace_id_(trace_id), span_id_(span_id) {}
+
+public:
+  std::unique_ptr<ProgressContext> Clone() const noexcept override;
```
"Copy"?

patrickkenney9801 replied:
I stole this from OpenTracing at:
https://github.com/opentracing/opentracing-cpp
but it is our API wrapper so we can change this if you want

ddn0 replied:
`Clone` is fine too.

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+  void Close() override {}
+};
+
+class KATANA_EXPORT JsonContext : public ProgressContext {
```
"JSONContext"

ddn0 commented on libsupport/include/katana/JsonTracer.h:
```diff
@@ -0,0 +1,98 @@
+  std::string span_id_;
+};
+
+class KATANA_EXPORT JsonSpan : public ProgressSpan {
```
"JSONSpan"

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+class ProgressSpan;
+class ProgressContext;
+
+// This class does not currently support thread-local tracers or concurrency controls
```
Probably write this as doxygen comment on the file itself. You can take a look at `Result.h` as an example.

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
+  static void SetProgressTracer(std::shared_ptr<ProgressTracer> tracer);
+
+  /// Create a new top level span if one does not already exist.
```
In general, 

"""
StartActiveSpan creates a new span. If a span does not already exist, create a new top-level span. Otherwise, this function creates a child span of the current span. By default, the returned scope will finish the span on close. To change this behavior, set finish_on_close to false.
"""

Also, as a matter of style, it is nice for options to default to false. So, it might be better for "finish_on_close" to be named something like "leave_open_on_close".

patrickkenney9801 replied:
this also comes from OpenTracing's Python API with finish_on_close=true, but we can change this of course

ddn0 replied:
Is the plan to eventually use the OpenTracing API directly later?

patrickkenney9801 replied:
Not directly, we will use some parts directly like its Value class
And then we will use OpenTracers like Jaeger or Lightstep wrapped in a subclass of ProgressTracer
Also OpenTracing's C++ API does not have a concept of "ProgressScope" like its Python API that is nicer to use

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+  /// Can specify to not finish the span when the scope goes out of scope
+  /// by RAII by setting finish_on_close=false
+  /// If called with a nullptr span,
+  ///    calls StartActiveSpan("error: nullptr span given to SetActiveSpan", finish_on_close)
```
Seems like a nullptr should just be a fatal error. What use-cases do we want to support by accepting a null span?

patrickkenney9801 replied:
Just to prevent having the tracing system throw an error, I can change it to fatal though if you think that's best

ddn0 replied:
It would be good to get to a point where nullptr is not a valid way to refer to a span/scope/etc. (if possible). This simplifies the life of users of the API.

patrickkenney9801 replied:
Alight I'll just make it throw an error then, the only way this could really happen is if a developer deliberately sends a nullptr

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+  /// Get the current scope’s span to add tagging/logging without having
+  /// to pass around scopes/spans as parameters
+  virtual std::shared_ptr<ProgressSpan> GetActiveSpan();
+  uint32_t GetHostId() { return host_id_; }
```
"GetHostID"

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of) = 0;
```
if we simply want to borrow a reference and not take ownership of a smart pointer, we should just take `const T&` or `T*` as appropriate.

patrickkenney9801 replied:
Are both this comment and the following referring to line 109, or is this for the unique_ptr?

ddn0 replied:
It is a general comment.

If we simply need to refer to an existing object for the duration of a call, `const T&` or `T*`, is more appropriate.

1. `Fn(T* t)` / `Fn(const T& t)`: caller promises to keep `t` live for the duration of the call, callee can access t for the duration of the call
2. `Fn(std::unique_ptr<T> t)`: caller is giving `t` to the callee
3. `Fn(const std::unique_ptr<T>& t)`: redundant with 1 and 3 precludes a `t` that is stack allocated.
4. `Fn(std::shared_ptr<T> t)` / `Fn(const std::shared_ptr<T>& t)`: caller and callee will share responsibility for deallocating `t`. Some people make a subtle distinction between the two forms (will share vs may share) [1], but as a matter of style, it might be simpler to just strengthen everything to `std::shared_ptr<T>`.

[1] https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

Not knowing the code, I don't know if it is always okay to replace `shared_ptr<T>` with `T*`. We should replace `const std::unique_ptr<T>&` with `const T&` at least.

patrickkenney9801 replied:
Okay I get it now, thanks

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, bool ignore_active_span = false) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
```
ditto

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+
+  /// Close is called when a tracer is finished processing spans
+  /// This should be called to ensure any and all buffered spans are
+  /// flushed
```
Do users need to call  `Close`  explicitly always or is this part of the interface with subclasses? Could `Close` be protected and called in the destructor?

patrickkenney9801 replied:
It's dependent on subclass implementations but it is required by Jaeger and Lightstep which I've been looking at so far, probably all mature tracers require it to behave correctly

I don't think it would be an issue to have this called in the destructor, but the tracer of a program will be stored in a static variable.  Not too sure about C++ destructor semantics w.r.t. this but if that's fine then I can just move it to the destructor

ddn0 replied:
There are two interacting issues.

1. How the object cleans itself up.
```cpp
// RAII
{
  Foo foo;
}

// Explicit management
{
  Foo foo;
  foo.Close();
}

// Hybrid (Close when out of scope and not already closed)
{
  Foo foo;
  //foo.Close();
}
```

2. How singletons are managed.

```cpp
// 1. Static variable 
Foo foo;

// 2. Static class variable
Foo Foo::foo;

// 3. Managed pointer
std::unique_ptr<Foo> foo;
```

The issue with both 1 and 2 is that when exactly `foo` is constructed and destructed is indeterminate in C++.  The issue with 3 is that you need to know when to construct (and probably destruct) foo, i.e., the construction must precede any uses and the destruction must be after the last use. This becomes difficult if there are references to foo in other static variable constructors.

I'd prefer 3 if possible because once we have to deal ordering of static initializers, it is hard to undo that.

patrickkenney9801 replied:
I messed something up with Github and had to post this below as well:

Right now we are using a static class variable shared_ptr to store the global tracer, eventually this will be thread local
Then we construct it in DistMemSys or SharedMemSys
And currently we have no destructor logic

ddn0 replied:
See https://github.com/KatanaGraph/katana/pull/370#discussion_r678598236 for some continued discussion of singletons.

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+
+  /// Get the scope’s underlying span to add tagging/logging,
+  /// span relationships, and handle multiple active spans at a time
+  std::shared_ptr<ProgressSpan> GetSpan() { return span_; }
```
can we return a reference rather than a shared_ptr (also maybe `span()`)?

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,225 @@
+public:
+  virtual ~ProgressContext() = default;
+  virtual std::unique_ptr<ProgressContext> Clone() const noexcept = 0;
+  virtual std::string GetTraceId() const noexcept;
```
"GetTraceID"

"GetSpanID"

ddn0 commented on libsupport/src/JsonTracer.cpp:
```diff
@@ -0,0 +1,391 @@
+// the tracing infrastructure is added and tested
+uint32_t id = 0;
+std::string
+GenerateId(uint32_t host_id) {
```
"GenerateID"

ddn0 commented on libsupport/src/JsonTracer.cpp:
```diff
@@ -0,0 +1,391 @@
+}
+
+std::string
+GetHostStatsJson() {
```
Json => JSON throughout

patrickkenney9801 reviewed (commented):
Right now we are using a static class variable shared_ptr to store the global tracer, eventually this will be thread local
Then we construct it in DistMemSys or SharedMemSys
And currently we have no destructor logic

tylershunt commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,233 @@
+public:
+  virtual ~ProgressTracer() = default;
+
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
```
Do we have to crate a new owner here? Could this be `ProgressTracer *` instead?

patrickkenney9801 replied:
To your above comment I still need to add a helper function to allow the distributed case and then I plan on adding this to import on top of the existing logging to make sure everyone agrees with it and serve as an example, then I will refactor the codebase and remove ProgressLogger
So this is still an experimental API and not supposed to be used by developers

That makes sense, but if we allow multiple tracers to be used by a program concurrently then it could be an issue
But we probably don't need that feature, and we could just say we don't allow it

patrickkenney9801 replied:
![image](https://user-images.githubusercontent.com/10097959/127376372-f32bee1f-47d4-40a5-b4be-93f49843189d.png)

tylershunt replied:
> That makes sense, but if we allow multiple tracers to be used by a program concurrently then it could be an issue
> But we probably don't need that feature, and we could just say we don't allow it

I don't understand what this has to do with copying the shared pointer? Or are you talking about something else?

patrickkenney9801 replied:
if the only shared_ptr copy of the tracer is the static variable, and we set the tracer to something else, then there would be no refs to it and it would be deleted right?

tylershunt replied:
Is having a single global pointer really the best way to implement multiple tracers? That seems confusing to me. If I'm setting the global tracer I would expect all users of the tracer to use the tracer I set. It seems like an anti-pattern to allow parts of the code to potentially hold a reference to an old tracer.

I would rather keep things simple.

patrickkenney9801 replied:
pushed a change switching it to a raw pointer, should be fine, is that what you had in mind?

ddn0 replied:
There are two points here:

First, returning or taking `std::share_ptr` should be looked at with some suspicion. If you mean to just provide access as long as the callee is live, `const T&`, `T*` and even `T&` for parameters is preferred:

All these forms,
```cpp
T& foo();
void Foo(const T&);
void Foo(T*);
```
signal to developers that `t` is borrowed and only needs to be live as long as the callee uses it.

When we write `std::shared_ptr` (or even `std::unique_ptr`) we are signaling to something about the management of the underlying resource. `std::shared_ptr` is particularly troublesome because it means you don't know who will use this resource and it may outlive the caller.

Second, if you want to implement a singleton, you can use this pattern:

```
class Foo {};

Foo& GetCurrentFoo();
void SetCurrentFoo(Foo); // or Foo&& 

static std::unique_ptr<Foo> foo;
```

If you really want, you can make `GetFoo` and `SetFoo` static methods on the class `Foo` instead of free functions, but the point here is (1) to not mix static members with static instances and (2) not allow callers to access the singleton except through `GetFoo` and `SetFoo`.

patrickkenney9801 replied:
is there an issue with the pyarrow dependency now or is that me?

patrickkenney9801 replied:
are the latest changes what you had in mind Donald?  I think that makes the most sense after talking to you and Tyler here

tylershunt reviewed (commented):
What I'm struggling with is envisioning what exactly this framework looks like when its; would it be possible to add an "example instrumentation" where you apply this to instrument some existing code?

That would  also provide a useful pattern for others to follow.

ddn0 commented on libsupport/include/katana/JSONTracer.h:
```diff
@@ -0,0 +1,84 @@
+      : ProgressTracer(host_id, num_hosts) {}
+
+public:
+  static std::shared_ptr<JSONTracer> Make(
```
shouldn't this return a `JSONTracer` or at the very least a `unique_ptr`?

ddn0 commented on libsupport/include/katana/JSONTracer.h:
```diff
@@ -0,0 +1,84 @@
+
+  JSONSpan(
+      const std::string& span_name,
+      const std::shared_ptr<ProgressSpan>& parent);
```
`const std::shared_ptr<ProgressSpan>&` is somewhat meaningless.

Either we want to borrow a reference, `const T&` or `T*` or if you really have to `T&`. Or you want to make a copy, `std::shared_ptr<T>`. `const std::shared_ptr<T>&` means you might make a copy or you might not and restricts callers to giving you a `shared_ptr`.

ddn0 commented on libsupport/include/katana/JSONTracer.h:
```diff
@@ -0,0 +1,84 @@
+      const std::string& span_name, bool ignore_active_span) override;
+  std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name,
+      const std::shared_ptr<ProgressSpan>& child_of) override;
```
Why do we need a class hierarchy over spans?

I can understand why we would want a variety of different output backends (JSON output, console output) but why would we need different implementations of a "progress span" or "progress context"? Any backend specific behavior can be predicated on the backend used not the span or context used.

patrickkenney9801 replied:
We'll wrap the OpenTracer Jaeger's span class with this for example, what do you mean by predicate?  Have a flag or callback?

ddn0 replied:
So "JSON{Tracer,Span,Context}" doesn't mean thing that produces JSON output but simply "tracing implementation that isn't OpenTracer Jaeger"? And the point is we want to support programs that have both OpenTracer Jaeger spans and our spans at the same time?

The point of a virtual class hierarchy is to allow extension and polymorphism within the same program.

If you just wanted abstraction, you can just as easily just use information hiding instead:

```cpp
class ProgressSpan {
  // Today OpenTracingSpan tomorrow WhizSpan
  OpenTracingSpan span_;

public: 
  int GetSpanID() const;
};
```

And the fact that there are a series of linked datatypes, `JSON{Tracer,Span,Context}` seems weird to me. What a span or logging context is should be independent of how it is rendered. If the idea is to support multiple implementations simultaneously, I can sorta see a reason for this but I still don't see why it couldn't just be `JSONTracer` `Span` and `Context`.

patrickkenney9801 replied:
My viewpoint was that we could use the JSONTracer for local development like through mpirun so we don't have to setup a Jaeger or other OpenTracing collector just to get logs while testing
Then for graph-workers we could have end-to-end tracing through a real OpenTracing implementation like Jaeger, maybe making it optional to get output to stdout through the JSONTracer stuff as well for just the graph-worker

I guess I don't really understand what you mean by a span or logging context should be independent of how it is rendered?  Maybe a better name would be MockTracer instead of JSONTracer?  It's a different tracing system than Jaeger regardless of how it is outputted?

ddn0 replied:
My point, boiled down, is that spans and contexts are part of the API of tracing. They say things like "this is a region of interest", "this region is contained in another region". Whether we have a mock backend or a real one, the regions are the same and their interpretation is the same. So, I don't (yet) understand why we would have different types of spans related to what backend we use.

To put it another way, why couldn't the mock backend and the open tracing backend take in the same definition of a span?

patrickkenney9801 replied:
Those are abstract interfaces in OpenTracing's API as well
https://github.com/opentracing/opentracing-cpp/blob/master/include/opentracing/span.h

patrickkenney9801 replied:
I guess another way of looking at it is:
As far as I know, no OpenTracing implementation actually builds a complete span and then sends it
What they do is send intermediary span data (like when you start/finish/log/add a tag) to a collector
Then the collector assembles the intermediary span data into complete spans and traces

My implementation does the same, that's why I added a python script to assemble the tracer output into a Jaeger compatible format

patrickkenney9801 replied:
So spans are an abstract concept and their implementations vary widely, so they need a common interface

patrickkenney9801 replied:
For contexts, those follow the same basic reasoning since OpenTracing allows "baggage items" that are propagated to every child span once they are added.  Then there is a description language for how to handle adding to baggage items if say "sum" is already present on it, then setting "sum" to 42 would add 42 to the current sum, or if it is a string maybe you want to concatenate
Then there is also the concept of dropping baggage items if they become too large since these need to be propagated to every child span so there is logic there as well

patrickkenney9801 replied:
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/src/jaegertracing/Span.h

patrickkenney9801 replied:
Actually I was wrong about saying all tracer implementations output intermediary span details
Looks like Jaeger at least waits until it has a complete span to report it

ddn0 replied:
Ok.

> As far as I know, no OpenTracing implementation actually builds a complete span and then sends it
> What they do is send intermediary span data (like when you start/finish/log/add a tag) to a collector
> Then the collector assembles the intermediary span data into complete spans and traces

Is a collector here something like JSONTracer or something else? In any event, these different ways of aggregating data seem seem orthogonal to the fact that the output will be JSON, which is the source of my confusion.

It also seems to lead to hacks like friend, which do not play well with class hierarchies.

If the intent is mirror the OpenTracing API, this virtual span business seems fine, but there are several differences with the OpenTracing API. For example, the OpenTracing API doesn't have any of this shared_ptr business for `StartSpan`: https://github.com/opentracing/opentracing-cpp/blob/master/include/opentracing/tracer.h#L92

Can you summarize why we need these differences?

patrickkenney9801 replied:
Mostly because I added ProgressScope for RAII and allow getting the active span
To be honest I've been looking mostly at the python API at:
https://opentracing-python.readthedocs.io/en/latest/api.html
So I am a bit unfamiliar with the C++ API, looking at Jaeger's examples and from what I can tell, C++'s OpenTRacing API does not support getting the active span so any time you want to create a child span or add span data you need to pass along the span like in
https://github.com/jaegertracing/jaeger-client-cpp/blob/master/examples/App.cpp
In my API we do not need to pass anything to get the same hierarchy, like in the python API

A collector would be run on a separate host in a cluster and aggregate span data-storing it in a database maybe or just making it available to be viewed at a certain web address

patrickkenney9801 replied:
Would you prefer I just close this and use the OpenTracing API instead?

patrickkenney9801 replied:
The C++ API lacks a main part of the OpenTracing API that I am trying to provide by adding a concept of an activespan and scopes
https://opentracing.io/docs/overview/scopes-and-threading/

patrickkenney9801 committed:
logging: refactor tracing to only use friend classes to hide constructors

patrickkenney9801 committed:
logging: make tracing API more minimal to make it harder to misuse

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+  void MarkScopeClosed();
+  virtual bool ScopeClosed();
+  bool IsFinished() { return finished_; }
+  const std::shared_ptr<ProgressSpan>& GetParentSpan() { return parent_; }
```
`ProgressSpan* GetParentSpan() { return parent_.get(); }` ?

patrickkenney9801 replied:
We need the shared pointer for logic in the tracer class

ddn0 commented on libsupport/src/JSONTracer.cpp:
```diff
@@ -0,0 +1,385 @@
+}
+
+katana::JSONSpan::JSONSpan(
+    const std::string& span_name, std::shared_ptr<katana::ProgressSpan> parent)
```
why do we take a `shared_ptr<katana::ProgressSpan>` but never do anything with parent itself? We could equally take `ProgressSpan* parent`.

patrickkenney9801 replied:
It is passed and saved in the parent constructor of ProgressSpan

ddn0 commented on libsupport/src/JSONTracer.cpp:
```diff
@@ -0,0 +1,385 @@
+katana::JSONSpan::Make(
+    const std::string& span_name,
+    std::shared_ptr<katana::ProgressSpan> parent) {
+  return std::shared_ptr<JSONSpan>(new JSONSpan(span_name, parent));
```
why return a shared_ptr when we allocate a new `JSONSpan` and don't keep a reference ourselves?

patrickkenney9801 replied:
For consistency since spans can be accessed in various ways

ddn0 replied:
If you want to support ease of access, it is better to return the type that says the least about something. unique_ptrs can always be made into shared_ptrs but the converse is not true.

If you want to support a variety of allocation patterns (e.g., object caching), maybe shared_ptr would make sense here.

Do you want to support a variety of allocation patterns or ease of use?

patrickkenney9801 replied:
I'll try and scope out how it works but it's mostly ease of access:

All spans maintain a shared_ptr reference to their parent so it does not go out of scope before the child
All spans give the user code some form of access to the shared_ptr-either directly or via a ProgressScope
The currently active span allows returning a reference to the object via a shared_ptr and passing the shared_ptr itself as the parent to a newly created child span

ddn0 replied:
Also, `std::move(parent)`

roshandathathri commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+///
+/// Best Practices:
+///   If possible always avoid creating raw ProgressSpans from ProgressTracer
+///   If possible always use ProgressScopes to handle ProgressSpans
```
What is the difference b/w tracer, scope, and span?

patrickkenney9801 replied:
These are from the OpenTracing specifications which can be found here: https://opentracing.io/docs/overview/spans/

In short spans are units of work with a defined start and stop point which we can add timestamped logs to as well as tags for the overall span (or unit of work)

Scopes are used to manage spans by being immovable and not copyable, and closing on RAII mechanics.  Whenever possible raw spans should not be used and scopes instead

Tracers are used to control span logic, they create spans and maintain the active span for ease of use for the developer

patrickkenney9801 replied:
Each Scope is "in charge of" a span for its lifetime, and closes it when it goes out of scope or the developer calls Close on it

roshandathathri replied:
Thanks Patrick. Please add this into the doxygen comments before you describe best practices.

roshandathathri commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+///      auto scope3 = tracer.StartActiveSpan("3");
+///   }
+///   Always results in scope2's ProgressSpan finishing before scope1's
+///   and scope3's ProgressSpan finishing before scope1's if there is no error
```
Couldn't parse this statement

patrickkenney9801 replied:
I'm trying to convey that you always get the behavior:

scope1 START
scope2 START
scope2 FINISH
scope1 FINISH

or

scope1 START
scope2 START
scope3 START
scope3 FINISH
scope2 FINISH
scope1 FINISH

depending on the error, as opposed to say scope1 finishing before scope2 so that scope2 is not contained entirely within the span of scope1

roshandathathri replied:
Change to this maybe?
"Always scope2's ProgressSpan finish before scope1's ProgressSpan; 
 if there is no error, scope3's ProgressSpan finish before scope1's ProgressSpan"

patrickkenney9801 replied:
That's much better thanks

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+};
+
+class KATANA_EXPORT ProgressSpan {
+  virtual void Close() = 0;
```
This is a pure virtual private method and JSONSpan actually makes this public. Should this be at least protected?

patrickkenney9801 replied:
That's a bug actually, thanks for spotting it, I'll make it private

ddn0 commented on libsupport/include/katana/JSONTracer.h:
```diff
@@ -0,0 +1,82 @@
+namespace katana {
+
+class KATANA_EXPORT JSONTracer : public ProgressTracer {
+  JSONTracer(uint32_t host_id = 0, uint32_t num_hosts = 1)
```
nit: as a matter of style, private methods go after public ones, e.g.,

```cpp
class Foo {
public:

private:
}
```

https://google.github.io/styleguide/cppguide.html#Declaration_Order

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+class KATANA_EXPORT ProgressScope {
+  friend class ProgressTracer;
+
+  ProgressScope(std::shared_ptr<ProgressSpan> span) : span_(span) {}
```
```ProgressScope(std::shared_ptr<ProgressSpan> span): span_(std::move(span)) {}```

There are a few places in this PR where we can avoid extra copies of shared_ptrs

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
```
Since this is a virtual base class, probably delete move and copy special member functions.

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+  virtual void Close() = 0;
+
+protected:
+  ProgressSpan(std::shared_ptr<ProgressSpan> parent) : parent_(parent) {}
```
`parent_(std::move(parent))`

ddn0 commented on libsupport/include/katana/ProgressTracer.h:
```diff
@@ -0,0 +1,222 @@
+public:
+  /// If Finish has not already been called for the ProgressSpan, it's
+  /// destructor must do so.
+  virtual ~ProgressSpan() = default;
```
probably delete move and copy member functions.

patrickkenney9801 replied:
To be clear for these you want me to do something like:
  ProgressTracer(const ProgressTracer&) = delete;
  ProgressTracer(ProgressTracer&&) = delete;
  ProgressTracer& operator=(const ProgressTracer&) = delete;
  ProgressTracer& operator=(ProgressTracer&&) = delete;
  
and then reset them to default in the child classes like:
  JSONTracer(const JSONTracer&) = default;
  JSONTracer(JSONTracer&&) = default;
  JSONTracer& operator=(const JSONTracer&) = default;
  JSONTracer& operator=(JSONTracer&&) = default;

patrickkenney9801 replied:
clang doesn't like this so I am going to remove it

ddn0 replied:
I'm not sure why you'd need to reset them in the child classes. `= default` (confusingly) means "compiler, synthesize the function if you can". In the case where your parent class is not movable/copyable, the "default" implementation for a child class is not viable:

https://godbolt.org/z/Kj9xrz9n6

and it would be clearer to omit them altogether, which is sort of the point of deleting the special member functions in the base class to begin with. If we intend to access objects through their base class, we probably shouldn't let objects be copied otherwise objects will be sliced.

Is it the case that we need JSONTracer to be copyable? If so, we'd have to define the special member functions explicitly, though I would ask under what circumstances we need a Tracer to be copyable, as that seems to go against the pointer based API we have now.

ddn0 commented on libsupport/src/JSONTracer.cpp:
```diff
@@ -0,0 +1,385 @@
+katana::JSONTracer::StartSpan(
+    const std::string& span_name,
+    std::shared_ptr<katana::ProgressSpan> child_of) {
+  return JSONSpan::Make(span_name, child_of);
```
`std::move(child_of)`

ddn0 reviewed (approved):
mostly nits.

I have some preference for `Make` functions to return `unique_ptr`s, otherwise it makes it look like ProgressSpans have some immutable/copy-on-write behavior but they really are mutable objects that are allocated and then destroyed.
